util/util.go

```diff
@@ -29,16 +29,75 @@ func CreateLog(host peer.ID) (info thread.LogInfo, err error) {
 	if err != nil {
 		return
 	}
-	addr, err := ma.NewMultiaddr(fmt.Sprintf("/p2p/%s", host.String()))
+	pro := ma.ProtocolWithCode(ma.P_P2P).Name
+	addr, err := ma.NewMultiaddr("/" + pro + "/" + host.String())
 	if err != nil {
 		return
 	}
 	return thread.LogInfo{
 		ID:        id,
 		PubKey:    pk,
 		PrivKey:   sk,
-		ReadKey:   rk.Bytes(),
-		FollowKey: fk.Bytes(),
+		ReadKey:   rk,
+		FollowKey: fk,
 		Addrs:     []ma.Multiaddr{addr},
 	}, nil
 }
+
+// GetOrCreateLog returns the log with the given thread and log id.
+// If no log exists, a new one is created under the given thread.
+func GetOrCreateLog(t tserv.Threadservice, id thread.ID, lid peer.ID) (info thread.LogInfo, err error) {
+	info, err = t.Store().LogInfo(id, lid)
+	if err != nil {
+		return
+	}
+	if info.PubKey != nil {
+		return
+	}
+	info, err = CreateLog(t.Host().ID())
+	if err != nil {
+		return
+	}
+	err = t.Store().AddLog(id, info)
+	return
+}
+
+// GetOwnLoad returns the log owned by the host under the given thread.
+func GetOwnLog(t tserv.Threadservice, id thread.ID) (info thread.LogInfo, err error) {
+	logs, err := t.Store().LogsWithKeys(id)
+	if err != nil {
+		return info, err
+	}
+	for _, lid := range logs {
+		sk, err := t.Store().PrivKey(id, lid)
+		if err != nil {
+			return info, err
+		}
+		if sk != nil {
+			li, err := t.Store().LogInfo(id, lid)
+			if err != nil {
+				return info, err
+			}
+			return li, nil
+		}
+	}
+	return info, nil
+}
```

Comment:
Sugg: unname `thread.LogInfo` param? or maybe assign in the non-case error (instead of `li`). Got some seconds thinking about when `info` was assigned :smile:

Reply:
cool, I cleaned up the non-error case for readability